ceilometermiddleware/swift.py

```diff
@@ -58,6 +58,11 @@
     password = a_big_secret
     interface = public
 """
+
+import eventlet
+eventlet.sleep()
+eventlet.monkey_patch()
```

Comment:
@manuvakery1 Better to put this in `__init__.py` because, the function calls comes between the imports which may not be standard practice.

Reply:
@manuvakery1 @vikram-idrive This should go into `Swift.py` as thats where this code is important. If `Swift.py` is loaded directly, this will not work as `__init__.py` will not be loaded unless you load the `ceilometer` module. In general, code specific to a module should be within the module, not in global context.

Please include me in all core project code reviews even if I am not required as an approving reviewer to deliver the code fast. I want to be in the loop for critical fixes like these.

Reply:
@diganta-dutta 
1. In our `proxy-server.conf`, we use `paste.filter_factory = ceilometermiddleware.swift:filter_factory`.  i.e. `ceilometermiddleware.swift` which is accessing it as a package. So `__init__.py` will always executed. I do not think this will change at all. So this is fully under our control. 
2. Also above code is no where being used in the `Swift.py` and so it's not directly related to it.
3.  Also using it in `Swift.py` violates [pep8](https://www.python.org/dev/peps/pep-0008/#imports)(point 2).

So considering all of these, i think `__init__.py` is the right place for the above code

Reply:
@vikram-idrive for 1. and 2., the code is related to the rest of the code in swift.py coz without it the ceilometermiddleware doesn’t work properly. So i feel it should go there as it would be much easier to understand in the same context. In `__init__.py` typically we don’t put code, only imports as in my personal experience i have seen we don’t check that file typically for a module’s logic. Newer python (3.3 i think) has even removed it I think. And about using it in our code, yes specifically in this case we are using as a module. But i am talking about a general coding guideline. If a python file is directly executed, this code being out will not be executed. init file according to me should only have code, if any, which relates to the module being loaded.

I didn’t get the pep8 violation. Can you explain?

Reply:
@diganta-dutta  Writing the code `__init__py` is used very often mostly used for common configurations for that package or some kind of initialization. Check out the following, how it's being used:
- [swift](https://github.com/openstack/swift/blob/master/swift/__init__.py)
- [swift's crypto middleware](https://github.com/openstack/swift/blob/master/swift/common/middleware/crypto/__init__.py)
- [gnocchi](https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/storage/__init__.py)
- [keystone](https://github.com/openstack/keystone/blob/master/keystone/server/__init__.py)
- [request library](https://github.com/psf/requests/blob/master/requests/__init__.py).

Newer python 3.3. removed it only for [Namespace packages](https://docs.python.org/3/reference/import.html#packages) i.e. to just create a package, empty `__init__.py` files can be omitted. Check this related [PEP 420](https://www.python.org/dev/peps/pep-0420/). Also check [this](https://stackoverflow.com/a/48804718/7455759) and [this](https://stackoverflow.com/a/448279/7455759) as well.

Since the above monkey patching is some kind of initialization for this package, i preferred to put it in `__init__.py`.

Regarding the `pep8` violation, the code:
```
eventlet.sleep()
eventlet.monkey_patch()
```
are function calls which if we put in on the top, the other imports comes after that and thus using it in that way is will give us the following [pep8 error](https://pep8.readthedocs.io/en/latest/intro.html#error-codes):
```
E402 | module level import not at top of file
```

Reply:
@diganta-dutta can i proceed with this or should i move it to swift.py . once this is finalized i can raise PR for openstack project to update the middleware to new release

Reply:
@vikram-idrive I still feel since the eventlet code changes the behavior of the code inside `Swift.py`, it should go there. However, I found [this guideline](https://specs.openstack.org/openstack/openstack-specs/specs/eventlet-best-practices.html) which recommends using it in `__init__.py`. So we can leave it as it is. @manuvakery1 Please ensure all the other points mentioned in the above url are considered.